ext/standard/string.c

```diff
@@ -2389,16 +2401,22 @@ PHP_FUNCTION(substr)
 {
 	zend_string *str;
 	zend_long l = 0, f;
+	zend_bool l_is_default = 1;
 	int argc = ZEND_NUM_ARGS();
 
 	ZEND_PARSE_PARAMETERS_START(2, 3)
 		Z_PARAM_STR(str)
 		Z_PARAM_LONG(f)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG(l)
+		Z_PARAM_LONG_EX(l, l_is_default, 1, 0)
 	ZEND_PARSE_PARAMETERS_END();
 
+	if (l_is_default) {
+		php_error_docref(NULL, E_DEPRECATED, "Method will produce a different result in PHP 8 when length is null");
+	}
+
 	if (f > (zend_long)ZSTR_LEN(str)) {
+		php_error_docref(NULL, E_DEPRECATED, "Method will return empty string for out-of-bound offsets in PHP 8");
```

Comment:
We could return the offset value and the string length here for supplemental information.